src/dimse/wadoRs.ts

```diff
@@ -1,72 +1,282 @@
+  const logger = LoggerSingleton.Instance;
+  // Set up all the paths and query levels.
+  const storagePath = config.get(ConfParams.STORAGE_PATH) as string;
+  let queryLevel = QUERY_LEVEL.STUDY
+  const studyPath = path.join(storagePath, studyInstanceUid);
+  let pathname = studyPath;
+  let filename = '';
+  if (seriesInstanceUid) {
+    queryLevel = QUERY_LEVEL.SERIES
+  }
+  if (sopInstanceUid) {
+    filename = sopInstanceUid;
+    pathname = path.join(pathname, sopInstanceUid);
+  }
 
+  // Is the path that we have a directory or a file?
+  let isDir = true;
+  if (await fileExists(pathname)) {
+    const stat = await fs.stat(pathname);
+    isDir = await stat.isDirectory();
+  }
+  let useCache = false;
+  const foundInstances: InstanceInfo[] = [];
+  if (isDir) {
+    // It's a directory, what things do we expect to find in this directory for this search?
+    const json = deepmerge.all(await doFind(QUERY_LEVEL.IMAGE, { StudyInstanceUID: studyInstanceUid, SeriesInstanceUID: seriesInstanceUid ?? '', SOPInstanceUID: sopInstanceUid ?? '' }), { arrayMerge: combineMerge}) as QidoResponse[];
+    const foundPromises = await Promise.all(json.map(async (instance) => {
+      if (instance['00080018']) {
+        const instanceUid = instance['00080018'].Value[0];
+        const seriesUid = instance['0020000E'].Value[0];
+        foundInstances.push({
+          study: studyInstanceUid,
+          series: seriesUid,
+          instance: instanceUid
+        });
+        return fileExists(path.join(studyPath, instanceUid));
+      }
+      return true;
+    }));
+
+    // If all of the files for this search exist, then we're gonna use the cache!
+    useCache = foundPromises.reduce((prev, curr) => prev && curr, true);
+  }
+  else {
+    // If the file exists, use the cache
+    useCache = await fileExists(pathname);
+  }
+
+  if (!useCache) {
+    // We're not using the cache, so go and fetch the files. This will happen even if just one file is missing.
+    // Could this be improved to just get the needed files?
+    logger.info(`fetching ${pathname}`);
+    await waitOrFetchData(studyInstanceUid, seriesInstanceUid ?? '', sopInstanceUid ?? '', queryLevel);
+  }
+
+  // We only need a thumbnail - get it and bail.
+  if (dataFormat === 'thumbnail') {
+    // Just use the first of the foundInstances for this search
+    const filePath = isDir ? path.join(pathname, foundInstances[0].instance as string) : pathname;
+    const buff = await convertToJpeg(filePath, true);
+    if (buff) {
+      return {
+        contentType: 'image/jpeg',
+        buffer: buff
+      };
+    }
+    else {
+      throw new Error('Failed to create thumbnail');
+    }
+  }
+
+  let buffers: (Buffer | undefined)[] = [];
+  try {
+    if (isDir) {
+      // We're in a directory, loop through the files we want and attach them to the return buffer
+      const files = await fs.readdir(pathname)
+      buffers = await Promise.all(files.map(async (file) => {
```

Comment:
don't *think* this needs to be async.

Is there ANY chance that we could ask for same file twice?  If so, thinking it would be simpler to just have a for loop with await - you'd only add to buffers array sequentially, but you'd stop the issue of attempting 2 file descriptors on same file.

Reply:
There is no chance we can access the same file as we're looping through a list of files from the filesystem which we have just requested. Doing it this way means we can multitask. There is a reason our linting forbids await in a for loop...